src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs

```diff
@@ -40,6 +40,62 @@ public class NpgsqlMigrationsSqlGenerator : MigrationsSqlGenerator
                 ?? throw new InvalidOperationException("No string type mapping found");
         }
 
+        public override IReadOnlyList<MigrationCommand> Generate(
+            IReadOnlyList<MigrationOperation> operations,
+            IModel model = null)
+        {
+            var results = base.Generate(operations, model);
+
+            var seededGeneratedColumns = operations
+                .OfType<InsertDataOperation>()
+                .Select(o => new { o.Schema, o.Table })
+                .Distinct()
+                .Select(t => new
+                {
+                    t.Schema,
+                    t.Table,
+                    Columns = (FindEntityTypes(model, t.Schema, t.Table) ?? Enumerable.Empty<IEntityType>())
+                        .SelectMany(e => e.GetDeclaredProperties()
+                            .Where(p => p.GetValueGenerationStrategy() ==
+                                        NpgsqlValueGenerationStrategy.IdentityByDefaultColumn ||
+                                        p.GetValueGenerationStrategy() ==
+                                        NpgsqlValueGenerationStrategy.IdentityAlwaysColumn ||
+                                        p.GetValueGenerationStrategy() == NpgsqlValueGenerationStrategy.SerialColumn)
+                            .Select(p => p.GetColumnName()))
+                })
+                .SelectMany(t => t.Columns.Select(p => new
+                {
+                    t.Schema,
+                    t.Table,
+                    Column = p,
+                }))
+                .ToList();
```

Comment:
`ToArray` is better even if you don't care about perf in that code.

Reply:
Changed to ToArray though it should be practically the same no?

In any case, this is very very non-perf-sensitive code.

Reply:
`List<T>` is a reference type allocated on the heap and which wraps an array which is reference type two. So this means that you get useless object which just points to the actual collection. Plus `ToArray` and `ToList` are not the same anymore because each iterator can implement them differently because of `IListProvider<T>`.

Another difference is a way how iteration over collections works. `foreach` over an array is translated into `for` without bound checks, but `foreach` over `List<T>` uses `List<T>.Enumerator` which checks the version and has bound checks.

Yeah, but what am doing is teaching you about internals (: